src/mysql.erl

```diff
@@ -389,49 +395,46 @@ transaction(Conn, Fun, Args, Retries) when is_list(Args),
-                _ ->
-                    %% Re-throw with the same trace. We'll use that in the
-                    %% final {aborted, {Reason, Trace}} in the outer level.
-                    erlang:raise(throw, {implicit_rollback, N - 1, Reason},
-                                 erlang:get_stacktrace())
-            end;
-        error:{implicit_commit, _Query} = E ->
+        case N > 0 of
+            true ->
+                case N of
+                    1 ->
+                        case Reason of
+                            {?ERROR_DEADLOCK, _, _} when Retries == infinity ->
+                                transaction(Conn, Fun, Args, infinity);
+                            {?ERROR_DEADLOCK, _, _} when Retries > 0 ->
+                                transaction(Conn, Fun, Args, Retries - 1);
+                            _OtherImplicitRollbackError ->
+                                %% This includes the case ?ERROR_LOCK_WAIT_TIMEOUT
+                                %% which we don't restart automatically.
+                                %% We issue a rollback here since MySQL doesn't
+                                %% seem to have fully rollbacked and an extra
+                                %% rollback doesn't hurt.
+                                ok = query(Conn, <<"ROLLBACK">>),
+                                {aborted, {Reason, Stacktrace}}
+                        end;
+                    _ ->
+                        %% Re-throw with the same trace. We'll use that in the
+                        %% final {aborted, {Reason, Trace}} in the outer level.
+                        erlang:raise(throw, {implicit_rollback, N - 1, Reason},
+                                     Stacktrace)
+                end;
+            false ->
+                handle_other_error(Conn, throw, {implicit_rollback, N, Reason}, Stacktrace)
+        end;
+        ?WITH_STACKTRACE(error, {implicit_commit, _Query} = E, Stacktrace)
             %% The called did something like ALTER TABLE which resulted in an
             %% implicit commit. The server has already committed. We need to
             %% jump out of N levels of transactions.
             %%
             %% Returning 'atomic' or 'aborted' would both be wrong. Raise an
             %% exception is the best we can do.
-            erlang:raise(error, E, erlang:get_stacktrace());
-        Class:Reason ->
-            %% We must be able to rollback. Otherwise let's crash.
-            ok = gen_server:call(Conn, rollback),
-            %% These forms for throw, error and exit mirror Mnesia's behaviour.
-            Aborted = case Class of
-                throw -> {throw, Reason};
-                error -> {Reason, erlang:get_stacktrace()};
-                exit  -> Reason
-            end,
-            {aborted, Aborted}
+            erlang:raise(error, E, Stacktrace);
+        ?WITH_STACKTRACE(Class, Reason, Stacktrace)
+        handle_other_error(Conn, Class, Reason, Stacktrace)
```

Comment:
I would prefer not moving this code to a function in this PR. It doesn't seem to be necessary for OTP 21 in particular. Just do the necessary changes, i.e. the ?WITH_STACKTRACE macro. Style changes like this can be done and discussed in a separate PR.